lib/libutee/include/user_ta_header.h

```diff
@@ -68,13 +68,28 @@ struct ftrace_buf {
 	uint64_t begin_time[FTRACE_RETFUNC_DEPTH]; /* Timestamp */
 	uint64_t suspend_time;	/* Suspend timestamp */
 	uint32_t curr_size;	/* Size of ftrace buffer */
+	uint32_t curr_idx;	/* Current entry in the (circular) buffer */
 	uint32_t max_size;	/* Max allowed size of ftrace buffer */
 	uint32_t head_off;	/* Ftrace buffer header offset */
 	uint32_t buf_off;	/* Ftrace buffer offset */
 	bool syscall_trace_enabled; /* Some syscalls are never traced */
 	bool syscall_trace_suspended; /* By foreign interrupt or RPC */
 };
 
+/*
+ * Structure pushed to the ftrace circular buffer on entry and exit of any
+ * instrumented function.
+ * - When entering the function, @stack_depth is non zero and @val is the
+ *   function address
+ * - When returning, @stack_depth is zero and @val is the time spent in the
+ *   function and its callees, in nanoseconds
+ */
+struct __ftrace_elem {
+	uint8_t stack_depth;
+	uint8_t reserved[7];
```

Comment:
This looks like wastage of space within ftrace buffer. Can't we encode stack depth in the top most nibble of function address?

Reply:
I tried to avoid making assumptions on the values of virtual addresses but it is reasonable to assume the topmost byte is zero, especially since ftrace and PAuth are currently incompatible. Not too sure about what values ASLR may bring but it can be disabled if need be.
I will push an update.